homeassistant/components/tolosauna/__init__.py

```diff
@@ -0,0 +1,113 @@
+"""Component to control TOLO Sauna/Steam Bath."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from typing import Any, cast
+
+from tololib import ToloClient
+from tololib.errors import ResponseTimedOutError
+from tololib.message_info import SettingsInfo, StatusInfo
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from ...helpers.entity import DeviceInfo
+from .const import DEFAULT_RETRY_COUNT, DEFAULT_RETRY_TIMEOUT, DOMAIN
+
+PLATFORMS = ["climate"]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up tolosauna from a config entry."""
+    client = ToloClient(entry.data[CONF_HOST])
+    coordinator = ToloSaunaUpdateCoordinator(hass, entry)
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {"client": client, "coordinator": coordinator}
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class ToloSaunaUpdateCoordinator(DataUpdateCoordinator):
+    """DataUpdateCoordinator for TOLO Sauna."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize ToloSaunaUpdateCoordinator."""
+        self._config_entry = entry
+        super().__init__(
+            hass=hass,
+            logger=_LOGGER,
+            name=f"{entry.title} ({entry.data[CONF_HOST]}) Data Update Coordinator",
+            update_method=self._update_data,
+            update_interval=timedelta(seconds=3),
+        )
+
+    async def _update_data(self) -> dict[str, Any]:
+        client = ToloClient(self._config_entry.data[CONF_HOST])
+        try:
+            status = client.get_status_info(
+                resend_timeout=DEFAULT_RETRY_TIMEOUT, retries=DEFAULT_RETRY_COUNT
+            )
+            settings = client.get_settings_info(
+                resend_timeout=DEFAULT_RETRY_TIMEOUT, retries=DEFAULT_RETRY_COUNT
+            )
+            return {"status": status, "settings": settings}
```

Comment:
If you used a NamedTuple or TypedDict or Dataclass as return type instead of a dict[str, Any] you could omit the casts in your CoordinatorEntity class `cast(StatusInfo, self.coordinator.data["status"])` and have the properties typechecked.

Reply:
I'm using a `NamedTuple` now (`ToloSaunaData`), but typing for `self.coordinator.data.status` still complains, so I went for `cast(StatusInfo, self.coordinator.data.status)`.

Reply:
the type of `ToloSaunaCoordinatorEntity.coordinator` is still `DataUpdateCoordinator`. Add `coordinator: ToloSaunaUpdateCoordinator` on line 88 https://github.com/home-assistant/core/blob/b006ba7ce36715221825692426782090ada9a7ed/homeassistant/components/tolosauna/__init__.py#L88 to override this, then you should be able to omit all `cast`.

Reply:
How can I get rid of this one?

https://github.com/home-assistant/core/blob/b006ba7ce36715221825692426782090ada9a7ed/homeassistant/components/tolosauna/__init__.py#L106-L111

Reply:
I don't think you can. Either keep the cast or add `# type: ignore[no-any-return]`.
Alternatively
```python
client: ToloClient = self.hass.data[DOMAIN][self._config_entry.entry_id]["client"] 
return client
```

Reply:
Ok, done, thx!

Reply:
Oh I may have found a way. You are setting up a client in the coordinator 
`client = ToloClient(self._config_entry.data[CONF_HOST])`
maybe it can be stored in an attribute (if it can be reused and doesn't have to be initialized for every request) in the coordinator. From the CoordinatorEntity you can then just reach for it `Self.coordinator.client`